### PR TITLE
Add upgrade guide for `auto_new_epoch`

### PR DESCRIPTION
**Pull request by kmaehashi**
Add upgrade guide for ``auto_new_epoch`` introduced in #4608.

### CODE REVIEW

File: docs/source/upgrade.rst [hunk 1]
```diff
@@ -8,6 +8,26 @@ This is a list of changes introduced in each release that users should be aware
 Most changes are carefully designed not to break existing code; however changes that may possibly break them are highlighted with a box.
 
 
+Chainer v5
+==========
+
+Changed Behavior of ``Optimizer.new_epoch``
+-------------------------------------------
+
+This change only affects to a minor number of users (who call ``new_epoch()`` method of Optimizers while using a trainer, or who implement their own Updater class).
```
Comment (niboshi) [inline docs/source/upgrade.rst]: * "This change should affect only a minority of users"
* ... ```:meth:`Optimizer.new_epoch` ``` method manually
* ```:class:`Updater` ``` class

File: docs/source/upgrade.rst [hunk 2]
```diff
@@ -8,6 +8,26 @@ This is a list of changes introduced in each release that users should be aware
...
+
+Optimizers provide ``new_epoch()`` method, which can be used to change optimizer behavior depending on the current epoch number.
```
Comment (niboshi) [inline docs/source/upgrade.rst]: optimizer behavior > the behavior of the optimizer

File: docs/source/upgrade.rst [hunk 3]
```diff
@@ -8,6 +8,26 @@ This is a list of changes introduced in each release that users should be aware
...
+Prior to Chainer v5, this method was expected to be called by users.
+In Chainer v5, Updaters call ``new_epoch()`` method of Optimizers automatically.
```
Comment (niboshi) [inline docs/source/upgrade.rst]: * Updaters -> updaters (lower case)
* new_epoch() method of Optimizers -> ```:meth:`Optimizer.new_epoch` ```

File: docs/source/upgrade.rst [hunk 4]
```diff
@@ -8,6 +8,26 @@ This is a list of changes introduced in each release that users should be aware
...
+If you are calling ``new_epoch()`` method manually while using a trainer (Updater), you may need any of the following fixes:
```
Comment (niboshi) [inline docs/source/upgrade.rst]: (Updater) -> (or an updater)

File: docs/source/upgrade.rst [hunk 5]
```diff
@@ -8,6 +8,26 @@ This is a list of changes introduced in each release that users should be aware
...
+
+* Pass ``auto_new_epoch=False`` to the constructor of the Updater to stop ``new_epoch()`` being called automatically by Updaters.
```
Comment (niboshi) [inline docs/source/upgrade.rst]: * "Pass ... to `Updater.__init__` to stop..." (with a link to `Updater` class)
* ... from being called ...
* Updaters -> the updater

File: docs/source/upgrade.rst [hunk 6]
```diff
@@ -8,6 +8,26 @@ This is a list of changes introduced in each release that users should be aware
...
+* Avoid calling ``new_epoch()`` method manually.
+
+If you are implementing your own Updater class, you may need to update your code to automatically call ``new_epoch()``.
+See `#4608 <https://github.com/chainer/chainer/pull/4608>`__ for the detailed discussions.
```
Comment (niboshi) [inline docs/source/upgrade.rst]: "... for a more detailed discussion" is more natural.
Btw, actually #4608 does not include any discussion. Is it worth mentioning?

File: docs/source/upgrade.rst (hunk 5 above)
Comment (kmaehashi) [reply docs/source/upgrade.rst]: > "Pass ... to Updater.__init__ to stop..." (with a link to Updater class)

As `__init__` is not a part of Updater interface, I added link to `StandardUpdater` as an example.

File: docs/source/upgrade.rst (hunk 6 above)
Comment (kmaehashi) [reply docs/source/upgrade.rst]: There're no discussion, but updater implementors may find it useful to understand how to fix their own updaters.
Changed to:
> See `#4608 <https://github.com/chainer/chainer/pull/4608>`__ for more details.

**Comment by kmaehashi** [comment]
Thanks for review! I fixed accordingly.

File: docs/source/upgrade.rst (hunk 5 above)
Comment (niboshi) [reply docs/source/upgrade.rst]: "constructor" is not a correct term in Python.

File: docs/source/upgrade.rst (hunk 6 above)
Comment (niboshi) [reply docs/source/upgrade.rst]: I think the information in that PR is (and should be) written in this documentation.

Comment (kmaehashi) [reply docs/source/upgrade.rst]: I assume not many people implements own Updaters, so it's rather better to keep it simple.

File: docs/source/upgrade.rst [hunk 7]
```diff
@@ -8,6 +8,26 @@ This is a list of changes introduced in each release that users should be aware
...
+
+Optimizers provide :meth:`~chainer.Optimizer.new_epoch` method, which can be used to change the behavior of optimizers depending on the current epoch number.
+Prior to Chainer v5, this method was expected to be called by users.
+In Chainer v5, updaters has been changed to call :meth:`~chainer.Optimizer.new_epoch` automatically.
+If you are calling :meth:`~chainer.Optimizer.new_epoch` method manually while using a trainer (or an updater), you may need any of the following fixes:
+
+* Pass ``auto_new_epoch=False`` to the constructor of the updater (e.g., :class:`~chainer.training.updaters.StandardUpdater`) to stop :meth:`~chainer.Optimizer.new_epoch` from being called automatically by the updater.
+* Avoid calling :meth:`~chainer.Optimizer.new_epoch` method manually.
+
+If you are implementing your own Updater class, you may need to update your code to automatically call :meth:`~chainer.Optimizer.new_epoch`.
```
Comment (niboshi) [inline docs/source/upgrade.rst]: Updater -> updater (lower case) or ```:class:`~chainer.training.Updater` ```

**Comment by niboshi** [comment]
Also added some comments.

**Comment by kmaehashi** [comment]
Fixed accordingly.
As discussed offline, I changed the sentence from:

> See `#4608 <https://github.com/chainer/chainer/pull/4608>`__ for the detailed discussions.

to:

> (you can refer to the changes introduced in `#4608 <https://github.com/chainer/chainer/pull/4608>`__ to understand how to fix your updater).

File: docs/source/upgrade.rst [hunk 8]
```diff
@@ -8,6 +8,25 @@ This is a list of changes introduced in each release that users should be aware
 Most changes are carefully designed not to break existing code; however changes that may possibly break them are highlighted with a box.
 
 
+Chainer v5
+==========
+
+Updaters Call ``Optimizer.new_epoch``
```
Comment (niboshi) [inline docs/source/upgrade.rst]: How about "Updators Automatically Call ..."?

File: docs/source/upgrade.rst [hunk 9]
```diff
@@ -8,6 +8,25 @@ This is a list of changes introduced in each release that users should be aware
...
+-------------------------------------
+
+This change should affect only a minority of users (who call :meth:`chainer.Optimizer.new_epoch` while using a trainer, or who implement their own :class:`~chainer.training.Updater` class).
```
Comment (niboshi) [inline docs/source/upgrade.rst]: Remove "chainer." prefix?

File: docs/source/upgrade.rst [hunk 10]
```diff
@@ -8,6 +8,25 @@ This is a list of changes introduced in each release that users should be aware
...
+
+Optimizers provide :meth:`~chainer.Optimizer.new_epoch` method, which can be used to change the behavior of optimizers depending on the current epoch number.
+Prior to Chainer v5, this method was expected to be called by users.
+In Chainer v5, updaters has been changed to call :meth:`~chainer.Optimizer.new_epoch` automatically.
```
Comment (niboshi) [inline docs/source/upgrade.rst]: updaters have been

File: docs/source/upgrade.rst [hunk 11]
```diff
@@ -8,6 +8,25 @@ This is a list of changes introduced in each release that users should be aware
...
+If you are calling :meth:`~chainer.Optimizer.new_epoch` method manually while using a trainer (or an updater), you may need any of the following fixes:
```
Comment (niboshi) [inline docs/source/upgrade.rst]: "If you have been calling"?

File: docs/source/upgrade.rst [hunk 12]
```diff
@@ -8,6 +8,25 @@ This is a list of changes introduced in each release that users should be aware
...
+
+* Pass ``auto_new_epoch=False`` to the constructor of the updater (e.g., :class:`~chainer.training.updaters.StandardUpdater`) to stop :meth:`~chainer.Optimizer.new_epoch` from being called automatically by the updater.
+* Avoid calling :meth:`~chainer.Optimizer.new_epoch` method manually.
+
+If you are implementing your own :class:`~chainer.training.Updater` class, you may need to update your code to automatically call :meth:`~chainer.Optimizer.new_epoch` (you can refer to the changes introduced in `#4608 <https://github.com/chainer/chainer/pull/4608>`__ to understand how to fix your updater).
```
Comment (niboshi) [inline docs/source/upgrade.rst]: "If you implement"?

**Comment by kmaehashi** [comment]
Fixed.

**Comment by niboshi** [comment]
LGTM.

**Pull request merged**